monitoring/ping.go

```diff
@@ -0,0 +1,306 @@
+		c.NewSerfClient = agent.NewSerfClient
+	}
+	return nil
+}
+
+// NewPingChecker returns a checker that verifies accessibility of nodes in the cluster by exchanging ping requests
+func NewPingChecker(conf PingCheckerConfig) (c health.Checker, err error) {
+	err = conf.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	latencyTTLMap, err := ttlmap.New(latencyStatsCapacity)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := log.WithFields(log.Fields{trace.Component: "ping"})
+	logger.Debugf("using Serf IP: %v", conf.SerfRPCAddr)
+	logger.Debugf("using Serf Name: %v", conf.SerfMemberName)
+
+	client, err := conf.NewSerfClient(serf.Config{
+		Addr: conf.SerfRPCAddr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		logger.Debugf("node %s status %s", node.Name, node.Status)
+		if node.Status != pb.MemberStatus_Alive.String() {
+			continue
+		}
+		if node.Name == conf.SerfMemberName {
+			self = node
+			break // self node found, breaking out of the for loop
+		}
+	}
+	if self.Name == "" {
+		return nil, trace.NotFound("failed to find Serf member with name %s", conf.SerfMemberName)
+	}
+
+	return &pingChecker{
+		self:           self,
+		serfClient:     client,
+		serfRPCAddr:    conf.SerfRPCAddr,
+		serfMemberName: conf.SerfMemberName,
+		latencyStats:   *latencyTTLMap,
+		logger:         *logger,
+	}, nil
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	err := c.check(ctx, r)
+
+	if err != nil {
+		c.logger.Error(err.Error())
+		r.Add(NewProbeFromErr(c.Name(), "", err))
+		return
+	}
+	r.Add(&pb.Probe{Checker: c.Name(), Status: pb.Probe_Running})
+}
+
+// check runs the actual system status verification code and returns an error
+// in case issues arise in the process
+func (c *pingChecker) check(ctx context.Context, r health.Reporter) (err error) {
+
+	client := c.serfClient
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.checkNodesRTT(nodes, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// checkNodesRTT implements the bulk of the logic by checking the ping time
+// between this node (self) and the other Serf Cluster member nodes
+func (c *pingChecker) checkNodesRTT(nodes []serf.Member, client agent.SerfClient) error {
+	// ping each other node and fail in case the results are over a specified
+	// threshold
+	for _, node := range nodes {
+		// skip pinging self
```

Comment:
One more thing that occurred to me is that there may be failed/left members in the cluster which should deserve special treatment. At least, left members should not be considered. Also, I don't believe it makes much sense to include failed members in the round trip time computation.

Reply:
I added this additional check on line 130 and forgot to add it here too :man_facepalming: 

Fixed now.